models/cycle_gan_model.py

```diff
@@ -225,10 +240,46 @@ def optimize_parameters(self):
         # D_A and D_B
         self.set_requires_grad([self.netD_A, self.netD_B], True)
         self.set_requires_grad([self.netG_A, self.netG_B], False)
-        
-        self.compute_D_loss()      # calculate gradients for D_A and D_B
+
+        if self.opt.use_contrastive_loss_D:
+            self.compute_D_contrastive_loss()      # calculate gradients for D_A and D_B
+        else:
+            self.compute_D_loss()
         (self.loss_D/self.opt.iter_size).backward()
         
         self.compute_step(self.optimizer_D,self.loss_names_D)
                         
         self.niter = self.niter +1
+
+
+    def compute_D_contrastive_loss(self):
+        """Calculate contrastive GAN loss for the discriminator"""
+        fake_B = self.fake_B.detach()
+        fake_A = self.fake_A.detach()
+        
+        # Fake; stop backprop to the generator by detaching fake_B
+        pred_fake_A = self.netD_B(fake_A)
+        pred_fake_B = self.netD_A(fake_B)
+        # Real
+        pred_real_A = self.netD_B(self.real_A)
+        pred_real_B = self.netD_A(self.real_B)
+
+        nb_preds=int(torch.prod(torch.tensor(self.netD_A(self.real_B).shape)))
```

Comment:
Let's get the number of patches once. I believe it's a fixed number available from the options.